[14.0][ADD] purchase_manual_add_supplierinfo

chafique-delli opened the pull request:
This module allows from the purchase order, to complete manually the supplierinfo list of each product where the supplierinfo is missing.

chafique-delli committed:
[14.0][ADD] purchase_manual_add_supplierinfo

chafique-delli commented:
@sebastienbeau , @kev-roche

sebastienbeau commented on purchase_manual_add_supplierinfo/models/product.py:
```diff
@@ -0,0 +1,41 @@
+            return True
+
+    @api.model_create_multi
+    def create(self, vals_list):
```
TODO we need to fix it

sebastienbeau commented on purchase_manual_add_supplierinfo/views/purchase.xml:
```diff
@@ -0,0 +1,33 @@
+                <button
+                    name="create_missing_supplierinfo"
+                    type="object"
+                    icon="fa-th"
```
Use a better icon, we should use a icon that mean something for the end user.
also maybe it's better to put the icon at the end of the line ?

chafique-delli replied:
@sebastienbeau , it's done.

sebastienbeau commented on purchase_manual_add_supplierinfo/models/purchase.py:
```diff
@@ -0,0 +1,70 @@
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    all_supplierinfo_ok = fields.Boolean(string="supplierinfo is ok ?")
```
this should be a computed field (that depend on every line field)

chafique-delli replied:
@sebastienbeau , it's done.

sebastienbeau commented on purchase_manual_add_supplierinfo/models/purchase.py:
```diff
@@ -0,0 +1,70 @@
+
+    all_supplierinfo_ok = fields.Boolean(string="supplierinfo is ok ?")
+
+    def _check_supplierinfo(self):
```
The check method should call a check method on every line

sebastienbeau committed:
purchase_manual_add_supplierinfo: hack to be able to do an action without saving

For end user point of view, it great to have a button to confirm or skip
Sadly Odoo will save and then do the action.
A possibility will to unlink after saving, but this will create useless data
that can trigger recompute, job...
So the only solution is to hack the save so it do nothing and then save after
by calling the button "confirm"

sebastienbeau commented on purchase_manual_add_supplierinfo/models/product.py:
```diff
@@ -0,0 +1,63 @@
+# Copyright 2023 Akretion
```
can you rename the file "product.py" to "product_supplierinfo.py" ?
Naming is important

chafique-delli replied:
@sebastienbeau , it's done.

sebastienbeau commented on purchase_manual_add_supplierinfo/tests/test_purchase_manual_add_supplierinfo.py:
```diff
@@ -0,0 +1,50 @@
+        cls.product_7 = cls.env.ref("product.product_product_7")
+
+    def test_update_supplierinfo_from_purchase(self):
+        products = self.product_6 + self.product_7
+        products.write({"seller_ids": [(6, 0, [])]})
+        purchase_order = self._create_purchase_order(self.partner, products)
+        purchase_order.button_confirm()
```
This part of code can be moved into the setUpClass

chafique-delli replied:
@sebastienbeau , it's done.

sebastienbeau commented on purchase_manual_add_supplierinfo/tests/test_purchase_manual_add_supplierinfo.py:
```diff
@@ -0,0 +1,50 @@
+        products.write({"seller_ids": [(6, 0, [])]})
+        purchase_order = self._create_purchase_order(self.partner, products)
+        purchase_order.button_confirm()
+        self.assertEqual(purchase_order.all_supplierinfo_ok, False)
```
Never use 

 `self.assertEqual(purchase_order.all_supplierinfo_ok, False)`

But use

self.assertFalse(purchase_order.all_supplierinfo_ok)`

chafique-delli replied:
@sebastienbeau , it's done.

sebastienbeau commented on purchase_manual_add_supplierinfo/tests/test_purchase_manual_add_supplierinfo.py:
```diff
@@ -0,0 +1,50 @@
+            )
+            supplier = supplier_form.save()
+            action_window = supplier.update_from_purchase()
+        self.assertEqual(purchase_order.all_supplierinfo_ok, True)
```
same here

sebastienbeau commented on purchase_manual_add_supplierinfo/tests/test_purchase_manual_add_supplierinfo.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertEqual(purchase_order.all_supplierinfo_ok, True)
+
+    def test_update_supplierinfo_from_purchase_line(self):
+        products = self.product_6
+        products.write({"seller_ids": [(6, 0, [])]})
+        purchase_order = self._create_purchase_order(self.partner, products)
+        purchase_order.button_confirm()
```
not needed, if po create is in the setupclass

chafique-delli committed:
refactoring and improvements

sebastienbeau committed:
purchase_manual_add_supplierinfo: improve test and fix skiping from line, improve UI

sebastienbeau committed:
purchase_manual_add_supplierinfo_group: add module

sebastienbeau reviewed (approved):
LGTM (code review and test) I have done some minor change, and added a glue module to solve compatibility.

Now it's ready !! @PaulGoubert @Kev-Roche

sebastienbeau committed:
purchase_manual_add_supplierinfo: fix issue when creating supplier on template

sebastienbeau committed:
purchase_manual_add_supplierinfo: fix issue with note

HviorForgeFlow commented:
/ocabot merge nobump

HviorForgeFlow commented:
/ocabot merge nobump

dreispt commented:
Attempted rescue at #2720

HviorForgeFlow commented:
/ocabot merge nobump

dreispt commented:
Tests need fixing.

mourad-ehm commented:
/ocabot rebase
